### PR TITLE
Sparse logistic

### PR DESCRIPTION
**Pull request by pakozm**
- Addition of `sparse_logistic` activation function.
- Improvement of `autodiff` package to implement sparse auto-encoder.
- Addition of tests for comparison between different auto-encoders.

### CODE REVIEW

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 1]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+    buffer.printf("ann.components.actf.sparse_logistic{ name='%s', beta=%f, rho=%f  }", name.c_str(), this->beta, this->avg_act);
+    return buffer.to_string(buffer_list::NULL_TERMINATED);
+  }
+ 
+  void SparseLogisticActfANNComponent::multiplyDerivatives(MatrixFloat *input_units,
+						     MatrixFloat *output_units,
+						     MatrixFloat *input_errors,
+						     MatrixFloat *output_errors) {
+    UNUSED_VARIABLE(input_units);
+    Kernels::applyLogisticDerivative(output_errors, output_units);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Move this derivative to the end of the function, after the addition of sparsity penalty derivative.

File: packages/ann/ann/c_src/logistic_actf_component.h [hunk 2]
```diff
@@ -46,6 +46,26 @@ namespace ANN {
 
     virtual char *toLuaString();
   };
+
+  class SparseLogisticActfANNComponent: public LogisticActfANNComponent {
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.h]: Add here doxygen documentation with a reference to Andrew Ng document with the problem statement

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 3]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+
+    float eps = 1e-6;
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Add the library `#include "cmath_overloads.h"` and change this to:
```C++
float eps = AprilMath::Limits<float>::epsilon();
```

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 4]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+    // Computing average activations
+    MatrixFloat *current_avg = matSum(output_units, 0);
+    matScal(current_avg, 1.0f/static_cast<float>(output_units->getDimSize(0)));
+     
+    // -p/\hat{p} + (1-p)/(1-\hat{p}
+    // -p/\hat{p}
+    
+    MatrixFloat *sparse_errors = current_avg->clone();
+    matScalarAdd(sparse_errors, eps);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Could be better to use `matClamp` to avoid the zero problem?

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 5]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+    matDiv(sparse_errors, -avg_act);
+    
+    // (1-p)/(1-\hat{p})
+    MatrixFloat *aux = current_avg->clone();
+    matComplement(aux);
+    matScalarAdd(aux, eps);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Idem as before, Could be better to use `matClamp`?

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 6]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+    matDiv(aux, (1.0f-avg_act));
+    matAddition(sparse_errors, aux, sparse_errors);
+
+    matAbs(sparse_errors, sparse_errors);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: `matAbs` is supossed to be **in-place** if no `dest` argument is given, so better `matAbs(sparse_errors)`

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 7]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+
+    //Unfold
+    MatrixFloat *row = 0; 
+    for (int i = 0; i < output_units->getDimSize(0); ++i) {
+      row = output_errors->select(0,i, row);
+      matAxpy(row , beta, sparse_errors);
+    }
+  
+    delete row;
+    matCmul(output_errors, input_errors);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: I think it is necessary an auxiliary `error` matrix for the sparsity penalty derivatives, following this algorithm:
```
sparse_gradient = input_errors->clone();
sparse_gradient += sparse penalty derivative;
output_errors = logistic derivative;
matCmul(output_errors, sparse_gradients);
```

File: packages/ann/ann/c_src/logistic_actf_component.cc (hunk 1 above)
Comment (joapaspe) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: I'll move it to the end. So we could change it for the bp of the logistic function.

Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Be careful, instead of calling `backprop` method, you should call directly `multiplyDerivatives` method.

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 8]
```diff
@@ -58,4 +58,63 @@ namespace ANN {
...
+  }
+
+  char *SparseLogisticActfANNComponent::toLuaString() {
+    buffer_list buffer;
+    buffer.printf("ann.components.actf.sparse_logistic{ name='%s', beta=%f, rho=%f  }", name.c_str(), this->beta, this->avg_act);
+    return buffer.to_string(buffer_list::NULL_TERMINATED);
+  }
+ 
+  void SparseLogisticActfANNComponent::multiplyDerivatives(MatrixFloat *input_units,
+						     MatrixFloat *output_units,
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Indentation problem here, probably related with tabs instead of spaces ;-)

File: packages/ann/ann/c_src/logistic_actf_component.h [hunk 9]
```diff
@@ -46,6 +46,26 @@ namespace ANN {
...
+
+    private:
+       float beta;
+       float avg_act;
+    
+    protected:
+    virtual void multiplyDerivatives(Basics::MatrixFloat *input_units,
+				     Basics::MatrixFloat *output_units,
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.h]: Indentation problem ;-)

File: packages/ann/ann/c_src/logistic_actf_component.h [hunk 10]
```diff
@@ -46,6 +46,26 @@ namespace ANN {
 
     virtual char *toLuaString();
   };
+
+  class SparseLogisticActfANNComponent: public LogisticActfANNComponent {
+
+    private:
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.h]: Align private at the same level as class.

File: packages/ann/ann/c_src/logistic_actf_component.h [hunk 11]
```diff
@@ -46,6 +46,26 @@ namespace ANN {
...
+       float beta;
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.h]: The following had to be indented two spaces more than class keyword.
https://github.com/pakozm/april-ann/blob/master/STYLEGUIDE.md

File: packages/ann/ann/c_src/logistic_actf_component.h [hunk 12]
```diff
@@ -46,6 +46,26 @@ namespace ANN {
...
+       float avg_act;
+    
+    protected:
+    virtual void multiplyDerivatives(Basics::MatrixFloat *input_units,
+				     Basics::MatrixFloat *output_units,
+				     Basics::MatrixFloat *input_errors,
+				     Basics::MatrixFloat *output_errors);
+
+    public:
+        SparseLogisticActfANNComponent(const char *name, float beta, float avg_act);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.h]: Indent two spaces more than class keyword.

File: packages/ann/ann/c_src/logistic_actf_component.cc [hunk 13]
```diff
@@ -58,4 +59,76 @@ namespace ANN {
...
+    MatrixFloat *aux = current_avg; // reusing current_avg
+    matComplement(aux);
+    matClamp(aux, EPS, 1.0f);
+    matDiv(aux, (1.0f-target_avg_act));
+    
+    // -p/\hat{p} + (1-p)/(1-\hat{p} = sparse_gradients + aux
+    matAxpy(sparse_gradients, 1.0f, aux);
+    
+    // Normalize the error
+    // matScal(sparse_gradients, INV_DIM0);
```
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Do we need to normalize by the bunch size? I have a big doubt around this issue :S

File: packages/ann/ann/test/test-sparse.lua [hunk 14]
```diff
@@ -41,3 +41,58 @@ T("SparseDotProductTest",
       end
     end
 end)
+--
+T("SparseLogistic",
```
Comment (pakozm) [inline packages/ann/ann/test/test-sparse.lua]: I have added this test, which uses finite differences to check the gradients computed by sparse logistic component.

**Comment by pakozm** [comment]
It is needed to add the optional extra arguments for sdae pretraining algorithm, allowing to introduce sparse logistic activation function with its weight sparsity penalty and desired sparsity arguments.

File: packages/ann/ann/binding/bind_ann_base.lua.cc [hunk 15]
```diff
@@ -1379,6 +1379,28 @@ void lua_pushAuxANNComponent(lua_State *L, ANNComponent *value);
...
+  int argn = lua_gettop(L);
+  const char *name=0;
+  float beta = 1.0f;
+  float rho = 0.05f;
+  if (argn == 1) {
+    LUABIND_CHECK_PARAMETER(1, table);
+    check_table_fields(L, 1, "name", "beta", "rho", (const char *)0);
+    LUABIND_GET_TABLE_OPTIONAL_PARAMETER(1, name, string, name, 0);
+    LUABIND_GET_TABLE_OPTIONAL_PARAMETER(1, beta, float, beta, 1.0f); 
+    LUABIND_GET_TABLE_OPTIONAL_PARAMETER(1, rho, float, rho, 0.05f);
```
Comment (pakozm) [inline packages/ann/ann/binding/bind_ann_base.lua.cc]: I think it would be better to change in the binding the field `rho` by `desired_sparsity` and `beta` by `penalty_weight`, just because it is easier to understand the parameters in the binding. We can discuss better names.

Comment (joapaspe) [inline packages/ann/ann/binding/bind_ann_base.lua.cc]: I think we should use "sparsity" and "penalty". "Desired" is redundant and penalty_weight could be confused by weight_decay.

Other option is: desired_sparsity and penalty_sparsity but the sparsity is also redundant in both cases.

File: packages/ann/ann/binding/bind_ann_base.lua.cc [hunk 16]
```diff
@@ -1379,6 +1379,28 @@ void lua_pushAuxANNComponent(lua_State *L, ANNComponent *value);
...
+  int argn = lua_gettop(L);
+  const char *name=0;
+  float beta = 1.0f;
+  float rho = 0.05f;
+  if (argn == 1) {
+    LUABIND_CHECK_PARAMETER(1, table);
+    check_table_fields(L, 1, "name", "sparsity", "penalty", (const char *)0);
+    LUABIND_GET_TABLE_OPTIONAL_PARAMETER(1, name, string, name, 0);
+    LUABIND_GET_TABLE_OPTIONAL_PARAMETER(1, beta, float, beta, 1.0f); 
+    LUABIND_GET_TABLE_OPTIONAL_PARAMETER(1, rho, float, rho, 0.05f);
```
Comment (pakozm) [inline packages/ann/ann/binding/bind_ann_base.lua.cc]: Last two `LUABIND...` are wrong, the second argument is the Lua key in the table, and it has to be `penalty` and `sparsity` respectively.

File: packages/ann/ann/c_src/logistic_actf_component.cc (hunk 13 above)
Comment (pakozm) [inline packages/ann/ann/c_src/logistic_actf_component.cc]: Add a comment for the future: `// FIXME: Do we need to normalize by the bunch size?`

File: packages/ann/autoencoders/lua_src/stacked_denoising_autoencoder.lua [hunk 17]
```diff
@@ -100,7 +102,7 @@ function ann_autoencoders_ae:constructor(t)
 				     bias_weights="b-dec" .. params.index,
 				     transpose=true
 				   } ):
-    push( ann.components.actf[params.visible.actf]{ name="v-enc-" .. params.index} )
+    push( ann.components.actf[params.visible.actf](table.insert({ name="v-enc-" .. params.index}, params.hidden.params)))
```
Comment (pakozm) [inline packages/ann/autoencoders/lua_src/stacked_denoising_autoencoder.lua]: Should this `table.insert` be `table.merge`?

Comment (joapaspe) [inline packages/ann/autoencoders/lua_src/stacked_denoising_autoencoder.lua]: Solved.

File: packages/ann/ann/binding/bind_ann_base.lua.cc (hunk 16 above)
Comment (joapaspe) [inline packages/ann/ann/binding/bind_ann_base.lua.cc]: I solved this and also updated to_lua_string of sparsity logistic. In case that we had some trainer saved with rho and beta we should change rho and beta by sparsity and penalty in the persistence file. Otherwise it will fail.

**Comment by pakozm** [comment]
Ok, we wait for Travis CI, whenever it says ok, the code would be merged with devel branch. Nice! :-)

**Comment by pakozm** [comment]
Travis is sad :(
Check the error in the log:
https://s3.amazonaws.com/archive.travis-ci.org/jobs/46855794/log.txt

**Comment by joapaspe** [comment]
It doesn't seems complicated, I'll take a look. I have to update some functions in sdae layerwise pretraining, since the actf parameters must be propagated in the ae building function and now in some points are lost. Wait for that until merge

**Comment by pakozm** [comment]
Just one more thing, add new feature to changeslist file, and we will be ready :-)

**Comment by joapaspe** [comment]
I am doing experimentation with this branch and checking if there are bugs. I did small corrections. I think in one or two days we could merge.

**Comment by pakozm** [comment]
Ok, cool.

**Comment by pakozm** [comment]
Wait until tomorrow afternoon for merge, it is pending devel into master merge.
